File: 2020/02/haskell/Main.hs
module Main where

import qualified Data.Text as Text
import qualified Data.Text.IO as Text

type Line = (Int, Int, Char, String)

main :: IO ()
main = do
  input <- fmap Text.lines (Text.readFile "../../input/02.txt")
  let lines = map Text.unpack input
  let log = map (map getLog) lines
  let split = map words log
  let parsed = map parse split
  print $ length $ filter part1 parsed
  print $ length $ filter part2 parsed

getLog :: Char -> Char
getLog '-' = ' '
getLog ':' = ' '
getLog c = c

parse :: [String] -> Line
parse (a : b : c : d : _) = (read a, read b, head c, d)

part1 :: Line -> Bool
part1 (min, max, char, string) = (count >= min) && (count <= max)
  where
    count = length $ filter (== char) string

part2 :: Line -> Bool
part2 (min, max, char, string) = (fst == char) /= (snd == char)
  where
    fst = string !! (min - 1)
    snd = string !! (max - 1)

File: 2020/02/rust/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

struct Log {
    min: usize,
    max: usize,
    char: char,
    passowrd: String,
}

fn main() -> Result<()> {
    let file = File::open("../../input/02.txt")?;
    let buf = BufReader::new(file);
    let inputs: Vec<String> = buf.lines().map(|l| l.unwrap()).collect();
    let mut list = Vec::new();

    for input in inputs {
        let pair: Vec<&str> = input.split(' ').collect();
        let bound: Vec<&str> = pair[0].split('-').collect();
        let char: Vec<&str> = pair[1].split(':').collect();
        list.push(Log {
            min: bound[0].parse()?,
            max: bound[1].parse()?,
            char: char[0].chars().next().unwrap(),
            passowrd: pair[2].to_string(),
        });
    }

    let result = part1(&list);
    println!("{}", result);

    let result = part2(&list);
    println!("{}", result);

    Ok(())
}

fn part1(list: &[Log]) -> usize {
    let mut result = 0;
    for log in list {
        let mut count = 0;
        log.passowrd.chars().for_each(|c| if c == log.char { count += 1; });
        if count >= log.min && count <= log.max { result += 1; }
    }

    result
}

fn part2(list: &[Log]) -> usize {
    let mut result = 0;
    for log in list {
        let min = log.passowrd.chars().nth(log.min - 1).unwrap_or(' ');
        let max = log.passowrd.chars().nth(log.max - 1).unwrap_or(' ');
        if (log.char == min) ^ (log.char == max) { result += 1; }
    }

    result
}

